pageserver/src/restore_local_repo.rs

```diff
@@ -63,8 +66,21 @@ pub fn import_timeline_from_postgres_datadir(
             None => continue,
 
             // These special files appear in the snapshot, but are not needed by the page server
-            Some("pg_control") => continue,
-            Some("pg_filenode.map") => continue,
+            Some("pg_control") => import_nonrel_file(
+                timeline,
+                Lsn(0), // control file is not versioned
```

Comment:
Yepp, will do pg_control file versioned. But the main problem is checkpoint record.  It is also currently not versioned. And creating new version for each transaction commit (causing increment of nextXid) seems to be overkill. I will try to find some other solution, like storing version each N-th nextXid update.

Reply:
> In general, I'm very skeptical of trying to duplicate all the logic to keep the control file in sync with the changes in the compute node.

Do you have an idea how to implement it better? In any case, if we want to be able to spawn compute node at any moment in time, we must provide correct pg_control

Reply:
> > In general, I'm very skeptical of trying to duplicate all the logic to keep the control file in sync with the changes in the compute node.
> 
> Do you have an idea how to implement it better? In any case, if we want to be able to spawn compute node at any moment in time, we must provide correct pg_control


Not really. Except by letting Postgres handle all the non-relational records. Or we could modify Postgres to allow for starting up at an arbitrary point in time, without a checkpoint record. I think we'd still need to reconstruct most of the information that's currently stored in the control file, but maybe it would be better to do it explicitly. Then you wouldn't need to create the WAL segment with the synthetic checkpoint record in it. Then again, the code in StartupXLog() is very complicated too, so I'm not excited about modifying that either.

To recap, the control file consists of roughly four parts:

1. Fields that never change. That includes 'system_identifier', and all the compile-time constants like BLCKSZ. These are easy to handle.

2. Copies of various GUCs that affect hot standby. That includes 'wal_level', 'wal_log_hints', 'max_connections' and so on. Changes to these are WAL-logged with a XLOG_PARAMETER_CHANGE record. (checkpoint.full_page_writes is in that category too, but it's handled differently, for historical reasons I think). These are also pretty straightforward.

3. 'state', 'unloggedLSN', 'minRecoveryPoint' and other fields that need to be updated or preserved correctly across restarts. These are tricky, we need to have special logic for each of them, mirroring the similar logic in PostgreSQL itself. I guess the logic for Zenith is simpler though: state is always DB_SHUTDOWNED_IN_RECOVERY (right?), and minRecoveryPoint is the LSN that you're creating the pg_control file for.

4. The checkpoint struct. It is WAL-logged by checkpoint records. But if we try to build one from scratch, creating a "synthetic checkpoint" where one didn't exist in the original compute node, that's tricky. Then we need to track all the XID fields correctly.

All the code for 3. and 4. needs to be carefully commented, with good test coverage.

Reply:
There is still some non-rel WAL records which affect relations. For example list of dropped relations in commit/abort records. I am not sure that it is the only case but we have to handle it. I do not understand how it can be done with non-rel WAL replayed by Postgres without relations. Another think is tracking nextXid, which also can not be done without patching Postgres core.

The approach with reimplemnting our own redo handlers in Rust of non-rel WAL records  has only one disadvantage: duplication of code and risk of incompatibility with future Postgres releases. But as far as this parts of Postgres are expected to be quite conservative, I think that this risk is minimal (unless somebody will decide to completely eliminate SLRU from Postgres and replace it with shared buffers).

> state is always DB_SHUTDOWNED_IN_RECOVERY (right?)
Why? It seems to be that it should be DB_SHUTDOWNED.
It sems to me that we are trying to convince Postgres that it was normalyl shutdowned and no recovery is needed.

Reply:
> There is still some non-rel WAL records which affect relations. For example list of dropped relations in commit/abort records. I am not sure that it is the only case but we have to handle it. I do not understand how it can be done with non-rel WAL replayed by Postgres without relations. Another think is tracking nextXid, which also can not be done without patching Postgres core.

Yeah, we need to handle dropped relations in commit/abort records in any case.

> > state is always DB_SHUTDOWNED_IN_RECOVERY (right?)
> 
> Why? It seems to be that it should be DB_SHUTDOWNED.
> It seems to me that we are trying to convince Postgres that it was normalyl shutdowned and no recovery is needed.

Ok.